### PR TITLE
Additional Functions implementation for Eval Engine

### PR DESCRIPTION
**Pull request by yliuuuu**
## Relevant Issues
- [Closes/Related To] Issue #XXX

## Description
- exists
- cardinality
- bit_length
- octet_length
- Extract.
- overlay (rewrite)

## Other Information
- Updated Unreleased Section in CHANGELOG: **[YES/NO]**
  - < If NO, why? >

- Any backward-incompatible changes? **[YES/NO]**
  - < If YES, why? >
  - < For this purpose, we define backward-incompatible changes as changes that—when consumed—can potentially result in
errors for users that are using our public APIs or the entities that have `public` visibility in our code-base. >

- Any new external dependencies? **[YES/NO]**
  - < If YES, which ones and why? >
  - < In addition, please also mention any other alternatives you've considered and the reason they've been discarded >

- Do your changes comply with the [Contributing Guidelines](https://github.com/partiql/partiql-lang-kotlin/blob/main/CONTRIBUTING.md)
  and [Code Style Guidelines](https://github.com/partiql/partiql-lang-kotlin/blob/main/CODE_STYLE.md)? **[YES/NO]**

## License Information

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: partiql-spi/src/main/kotlin/org/partiql/spi/connector/sql/builtins/FnExtract.kt [hunk 1]
```diff
@@ -0,0 +1,368 @@
...
+            FnParameter("datetime", TIME),
+        ),
+        isNullCall = true,
+        isNullable = false,
+    )
+
+    override fun invoke(args: Array<PartiQLValue>): PartiQLValue {
+        val v = args[0].check<TimeValue>().value!!
+        return when (val tz = v.timeZone) {
+            TimeZone.UnknownTimeZone -> int32Value(0)
```
Comment (johnedquinn) [inline partiql-spi/src/main/kotlin/org/partiql/spi/connector/sql/builtins/FnExtract.kt:291]: QQ: You're the one most familiar with date time semantics, but should this be 0? While it's not a boolean value, in boolean logic, an unknown value gets coerced to a NULL value.

The Ion spec says:
> 2007-02-23T20:14:33.079Z         // The same instant in UTC ("zero" or "zulu")
2007-02-23T20:14:33.079+00:00    // The same instant, with explicit local offset
2007-02-23T20:14:33.079-00:00    // The same instant, with unknown local offset

Which shows that the timestamps themselves are equivalent. But, if you were to want to derive information from your data, I feel as if NULL would be returned upon extracting an unknown offset.

File: partiql-spi/src/main/kotlin/org/partiql/spi/connector/sql/exts/FnExists.kt [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+        isNullable = false,
+    )
+
+    override fun invoke(args: Array<PartiQLValue>): PartiQLValue {
+        val container = args[0].check<BagValue<*>>()
+        val iter = container.iterator()
+        if (iter.hasNext()) {
+            return boolValue(true)
+        }
+        return boolValue(false)
```
Comment (johnedquinn) [inline partiql-spi/src/main/kotlin/org/partiql/spi/connector/sql/exts/FnExists.kt:36]:
```suggestion
        val exists = container.iterator().hasNext()
        return boolValue(exists)
```

Nit -- non-blocking, but just a recommendation

File: partiql-spi/src/main/kotlin/org/partiql/spi/connector/sql/exts/FnSize.kt [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+        parameters = listOf(
+            FnParameter("container", PartiQLValueType.BAG),
+        ),
+        isNullCall = true,
+        isNullable = false,
+    )
+
+    override fun invoke(args: Array<PartiQLValue>): PartiQLValue {
+        val container = args[0].check<BagValue<*>>()
+        val iter = container.iterator()
```
Comment (johnedquinn) [inline partiql-spi/src/main/kotlin/org/partiql/spi/connector/sql/exts/FnSize.kt:32]: Same recommendations as CARDINALITY.

**Review by johnedquinn — COMMENTED** [review]
Overall looks great. Some minor comments.

File: partiql-spi/src/main/kotlin/org/partiql/spi/connector/sql/builtins/FnExtract.kt (hunk 1 above)
Comment (yliuuuu) [reply partiql-spi/src/main/kotlin/org/partiql/spi/connector/sql/builtins/FnExtract.kt:291]: This is a good question. 

While I don't know what should the decision be, for the time being I decided to follow the behavior from the existing pipelines, which returns zero (in form of decimal for some reasons)...

```
EXTRACT(TIMEZONE_HOUR FROM `2024-01-01T23:12:59-00:00`)
   | 
===' 
0.
```

I think for the time being we can stay with that behavior, but I will add a todo note for this line.

Great call out. Thanks.

**Review by johnedquinn — APPROVED** [review]
LGTM!

**Pull request merged**